File: ref/csharp/NelfeScoring.Vectors/Crypto.cs
using System.Security.Cryptography;
using System.Text;

namespace NelfeScoring;

/// <summary>
/// Primitives cryptographiques FIGÉES (SPEC v1.0 §5.1) :
///  - empreintes SHA-256 en hexadécimal minuscule ;
///  - signature device/serveur ECDSA P-256, clé publique en SPKI DER, signature en
///    ASN.1 DER (RFC 3279), le tout encodé base64url sans padding à l'extérieur ;
///  - key_id = SHA-256(SPKI_DER).
/// </summary>
public static class Crypto
{
    public static string Sha256Hex(byte[] data)
        => Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant();

    public static string Sha256Hex(string utf8) => Sha256Hex(Encoding.UTF8.GetBytes(utf8));

    public static string KeyId(byte[] spkiDer) => Sha256Hex(spkiDer);

    // ── base64url sans padding ────────────────────────────────────────────────
    public static string B64Url(byte[] data)
        => Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    public static byte[] FromB64Url(string s)
    {
        var t = s.Replace('-', '+').Replace('_', '/');
        switch (t.Length % 4) { case 2: t += "=="; break; case 3: t += "="; break; }
        return Convert.FromBase64String(t);
    }

    // ── signature (générateur) / vérification (CoreVerifier) ──────────────────
    public static string SignB64Url(ECDsa privateKey, byte[] message)
        => B64Url(privateKey.SignData(message, HashAlgorithmName.SHA256,
                                      DSASignatureFormat.Rfc3279DerSequence));

    /// <summary>Vérifie une signature base64url (ASN.1 DER) avec une clé SPKI DER.</summary>
    public static bool Verify(byte[] spkiDer, byte[] message, string signatureB64Url)
    {
        try
        {
            using var ec = ECDsa.Create();
            ec.ImportSubjectPublicKeyInfo(spkiDer, out _);
            return ec.VerifyData(message, FromB64Url(signatureB64Url),
                                 HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence);
        }
        catch { return false; }
    }

    /// <summary>Charge une clé privée EC depuis un PEM (pour le générateur de vecteurs).</summary>
    public static ECDsa LoadPrivate(string pemPath)
    {
        var ec = ECDsa.Create();
        ec.ImportFromPem(File.ReadAllText(pemPath));
        return ec;
    }

    /// <summary>Charge une clé et renvoie son SPKI DER (ce que le vérifieur manipule).</summary>
    public static byte[] LoadSpkiDer(string pemPath)
    {
        using var ec = ECDsa.Create();
        ec.ImportFromPem(File.ReadAllText(pemPath));
        return ec.ExportSubjectPublicKeyInfo();
    }
}

File: ref/php/src/Crypto.php
<?php
declare(strict_types=1);

namespace NelfeScoring;

/**
 * Primitives crypto (SPEC v1.0 §5.1). SHA-256 hex ; vérification ECDSA P-256 avec
 * clé publique SPKI et signature ASN.1 DER (ce que openssl attend nativement pour EC),
 * signature transportée en base64url ; key_id = SHA-256(SPKI DER).
 */
final class Crypto
{
    public static function sha256Hex(string $data): string
    {
        return hash('sha256', $data);
    }

    public static function fromB64Url(string $s): string
    {
        $t = strtr($s, '-_', '+/');
        $pad = strlen($t) % 4;
        if ($pad === 2) $t .= '==';
        elseif ($pad === 3) $t .= '=';
        return base64_decode($t, true) ?: '';
    }

    public static function b64url(string $data): string
    {
        return rtrim(strtr(base64_encode($data), '+/', '-_'), '=');
    }

    /**
     * Signe un message avec une clé privée EC P-256 (PEM). openssl_sign produit
     * nativement une signature ASN.1 DER pour EC — exactement ce que verify() attend.
     * @param string $privateKeyPem PEM (BEGIN EC PRIVATE KEY / PRIVATE KEY)
     */
    public static function signB64Url(string $privateKeyPem, string $message): string
    {
        $key = openssl_pkey_get_private($privateKeyPem);
        if ($key === false) return '';
        $signature = '';
        if (!openssl_sign($message, $signature, $key, OPENSSL_ALGO_SHA256)) return '';
        return self::b64url($signature);
    }

    /** @param string $publicKeyPem PEM SPKI (BEGIN PUBLIC KEY) */
    public static function verify(string $publicKeyPem, string $message, string $signatureB64Url): bool
    {
        $key = openssl_pkey_get_public($publicKeyPem);
        if ($key === false) return false;
        $der = self::fromB64Url($signatureB64Url);
        if ($der === '') return false;
        // openssl_verify : EC + SHA-256, signature en ASN.1 DER → 1 si valide.
        return openssl_verify($message, $der, $key, OPENSSL_ALGO_SHA256) === 1;
    }

    /** key_id = SHA-256(SPKI DER), à partir d'un PEM public. */
    public static function keyIdFromPubPem(string $publicKeyPem): string
    {
        $key = openssl_pkey_get_public($publicKeyPem);
        $details = openssl_pkey_get_details($key);
        // $details['key'] est le PEM SPKI ; on reconstruit le DER pour hacher.
        $der = self::pemToDer($details['key']);
        return hash('sha256', $der);
    }

    public static function pemToDer(string $pem): string
    {
        $body = preg_replace('/-----[^-]+-----|\s+/', '', $pem) ?? '';
        return base64_decode($body, true) ?: '';
    }
}
